proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,135 @@
+  - name: FileSimilarityRetriever
+    type: FileSimilarityRetriever
+    params:
+      document_store: DocumentStore
+      primary_retriever: EmbeddingRetriever
+      top_k: 30
+      file_aggregation_key: file_id
+      max_num_queries: 50
+
+pipelines:
+  - name: query
+    nodes:
+      - name: FileSimilarityRetriever
+        inputs: [Query]
+
+  - name: indexing
+    nodes:
+      - name: TextConverter
+        inputs: [File]
+
+      - name: EmbeddingRetriever
+        inputs: [TextConverter]
+
+      - name: DocumentStore
+        inputs: [EmbeddingRetriever]
+
+```
+
+# Motivation
+
+Initially this retriever was implemented by Mathis for a customer case to quickly retrieve similar files given a file as input. The reason we would like to add it to Haystack is because we see wider use for this node in future customer and community cases.
+
+One example where this could be useful is in academic writing, if you have a large number of scientific sources (journals, books, conference proceedings) stored and you'd like to find all that are similar to a specific article or to what you've written so far (in the latter case, it can also be helpful for plagiarism detection). And as for industry use cases, report writing in any field can be facilitated by file similarity retrieval. It can also be very helpful for lawyers preparing a case as well as journalists doing research for their articles.
+
+# Detailed design
+
+You can see the code for this proposal here: https://github.com/deepset-ai/haystack/pull/5666/files.
+
+As a general description, the FileSimilarityRetriever works by:
+1. Getting all documents corresponding to the provided file ID from the document store. The documents can be obtained together with their embeddings, to save resources by avoiding re-calculating them at query time.
+2. Retrieving similar docs to each document from the file, using one or two chosen retrieval methods.
+3.  a) Returning the top-k retrieved similar documents, if only one retrieval method was used.
+    b) Returning an aggregated list of similar documents retrieved by both retrievers, if two retrieval methods were chosen. The results are aggregated based on the reciprocal rank fusion score, though it is also possible to keep the original score (if one of the retrievers assigned the document a relevance score) in the metadata.
+    Reciprocal rank fusion is calculated as 1 / (k + index)
+    k is set to 61 in reciprocal rank fusion based on a University of Waterloo paper (co-authored with Google) called "Reciprocal Rank Fusion outperforms Condorcet and individual Rank Learning Methods" [https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf] where k=60 was suggested, and 1 was added as python lists are 0-based and the paper used 1-based ranking.
+4. Returning the top-k similar files, based on an aggregated relevance score of all documents pertaining to the file. In the results the user sees the most relevant document from each file.
```

Comment:
It isn't clear to me what top-k does. It seems like there are two top-k values that need to be set:
1. The number of documents to be returned in step 3
2. The number of files to be returned in step 4
They may need to be different if multiple documents from the same files. However, I don't know if it's realistic to guarantee a certain number of files because then we would need to set the first top-k to (number of splits in longest document) * (second top-k) which we would need to keep track of and which could be very big.

Reply:
I just noticed that this was already discussed. However, I think the proposal should probably be more clear on this.

Reply:
I've modified the description to (hopefully) make the top-k clearer - please let me know if it can be improved further 👍

Reply:
Thanks, that makes it clearer.